### ISSUE TITLE
Confusing goal with unfinished `have`

### ISSUE THREAD

**Issue by Vtec234** [leanprover/lean4#779]
### Description

In the following unfinished proof:
```lean
example : ∃n, n < 1 := by
   have h : 0 < 1 := 
                   --^ cursor here
   exact ⟨0, h⟩
```
the goal state is
```lean
h : 0 < 1
⊢ ∃ n, n < 1
```
with `0 < 1` as the assumption. To see the real goal we have to write `by`. This is confusing for beginners, since really at that cursor there is no goal state at all -- one should produce a *term*. So I suggest we show an `expected type` there rather than a goal state.

### Versions

8df2b07

**Comment by leodemoura** [comment]
@Vtec234 Note that the example above is currently parsed as
```lean
example : ∃n, n < 1 := by
   have h : 0 < 1 := exact ⟨0, h⟩
```
The `exact ⟨0, h⟩` is parsed as an application. The syntax tree is
```lean
(Term.app `exact [(Term.anonymousCtor "⟨" [(numLit "0") "," `h] "⟩")])
```
The info view shows the correct "Expected type" (`0 < 1`) when we hover over `exact`.

**Comment by leodemoura** [comment]
@Kha Any ideas?

**Comment by Kha** [comment]
Good question. If we can make it show the expected type here, do we want to hide the goal by default if we're inside of a term? We definitely don't want to always do that, e.g. when typing a `simp` argument.

**Comment by leodemoura** [comment]
I am inclined to mark this issue as "wontfix", and close it.

@Vtec234 I acknowledge the current behavior is confusing, but I don't see a good solution. Any suggestions on how to improve it?

**Comment by gebner** [comment]
I'm not sure why you expect the goal state to change here.  After `have h := ` comes a term, not a tactic.  (It would be great though if the expected type was shown directly after `:=`.)

Would it be less confusing if we had a `have h by ...` syntax?

**Comment by Vtec234** [comment]
> It would be great though if the expected type was shown directly after `:=`.

Yes, this is what I expected.

**Comment by gebner** [comment]
> Do you know why we are not getting the correct expected type here?

I'm pretty sure that's because we look for the smallest term node around the cursor, but there's no term node around the space after `:=` because the space is part of `:=`:

```lean
import Lean
open Lean Elab Command

elab "#dump" t:term : command => do
  logInfo (repr t)

#dump by have h :=
  exact

-- ^^ newline is part of :=
```

**Comment by Kha** [comment]
I don't think that would work if what follows the `:=` is either EOF or something that is definitely not a term (e.g. the next `theorem`). We could instead put an info node on `:=`, MVP for `let` only: https://github.com/Kha/lean4/commit/30278e8aeb50f3b508aaff98d4a0da1f0582951d
However, the output is still quite confusing given the unchanged outer goal 
![image](https://user-images.githubusercontent.com/109126/171825208-c89a81b3-6b5f-40b0-8888-d08cfdff86de.png)
